File: app/backend/classes/Quarto.py
import itertools

try:
    from classes.Hotel import Hotel
except ImportError:
    try:
        from app.backend.classes.Hotel import Hotel
    except ImportError:
        from Hotel import Hotel

class Quarto:
    id_counter = itertools.count(start=1)

    def __init__(
        self,
        id_hotel: int,
        tipo: str,
        status: str,
        andar: int,
        num_quarto: int,
        diaria: float,
        id_quarto: int = None,
        criado_em: str = None
    ) -> None:
        self.id = id_quarto if id_quarto is not None else next(Quarto.id_counter)
        self.id_quarto = self.id
        self.id_hotel = int(id_hotel)
        self.tipo = tipo
        self.status = status or "Disponível"
        self.andar = int(andar)
        self.num_quarto = int(num_quarto)
        self.diaria = float(diaria)
        self.criado_em = criado_em

    def get(self, varClass: str) -> int | float | str | None:
        match varClass:
            case "id" | "id_quarto": return self.id
            case "id_hotel": return self.id_hotel
            case "tipo": return self.tipo
            case "status": return self.status
            case "andar": return self.andar
            case "num_quarto": return self.num_quarto
            case "diaria": return self.diaria
            case "criado_em": return self.criado_em
            case _: return None

    def __set__(self, varClass: str, newValue: int | float | str):
        match varClass:
            case "id" | "id_quarto": self.id = self.id_quarto = int(newValue)
            case "id_hotel": self.id_hotel = int(newValue)
            case "tipo": self.tipo = str(newValue)
            case "status": self.status = str(newValue)
            case "andar": self.andar = int(newValue)
            case "num_quarto": self.num_quarto = int(newValue)
            case "diaria": self.diaria = float(newValue)
            case "criado_em": self.criado_em = str(newValue)
            case _: return ValueError

    def to_dict(self) -> dict:
        return {
            "id_quarto": self.id,
            "id": self.id,
            "id_hotel": self.id_hotel,
            "tipo": self.tipo,
            "status": self.status,
            "andar": self.andar,
            "num_quarto": self.num_quarto,
            "diaria": self.diaria,
            "criado_em": self.criado_em
        }

    @classmethod
    def from_dict(cls, data: dict):
        return cls(
            id_quarto=data.get("id_quarto") or data.get("id"),
            id_hotel=data.get("id_hotel", 1),
            tipo=data.get("tipo", "Casal"),
            status=data.get("status", "Disponível"),
            andar=data.get("andar", 1),
            num_quarto=data.get("num_quarto", 101),
            diaria=data.get("diaria", 150.0),
            criado_em=data.get("criado_em")
        )

File: app/frontend/src/models/Quarto.js
/**
 * Classe Quarto (Espelha Quarto.py do backend com métodos de domínio OO)
 */
export class Quarto {
  constructor({
    id = null,
    id_quarto = null,
    id_hotel = 1,
    tipo = 'Standard Casal',
    status = 'Disponível',
    andar = 1,
    num_quarto = 101,
    diaria = 150.0,
    criado_em = null
  } = {}) {
    this._id = id ?? id_quarto;
    this._idHotel = Number(id_hotel) || 1;
    this._tipo = tipo || 'Standard Casal';
    this._status = status || 'Disponível';
    this._andar = Number(andar) || 1;
    this._numQuarto = Number(num_quarto) || 101;
    this._diaria = Number(diaria) || 150.0;
    this._criadoEm = criado_em || new Date().toISOString();
  }

  get id() {
    return this._id;
  }

  set id(value) {
    this._id = value ? Number(value) : null;
  }

  get idHotel() {
    return this._idHotel;
  }

  set idHotel(value) {
    this._idHotel = Number(value);
  }

  get tipo() {
    return this._tipo;
  }

  set tipo(value) {
    this._tipo = String(value);
  }

  get status() {
    return this._status;
  }

  set status(value) {
    this._status = String(value);
  }

  get andar() {
    return this._andar;
  }

  set andar(value) {
    this._andar = Number(value);
  }

  get numQuarto() {
    return this._numQuarto;
  }

  set numQuarto(value) {
    this._numQuarto = Number(value);
  }

  get diaria() {
    return this._diaria;
  }

  set diaria(value) {
    this._diaria = Math.max(0, Number(value));
  }

  get criadoEm() {
    return this._criadoEm;
  }

  // Regras de Negócios e Predicados OO
  isDisponivel() {
    return this._status.toLowerCase() === 'disponível' || this._status.toLowerCase() === 'disponivel';
  }

  isOcupado() {
    return this._status.toLowerCase() === 'ocupado';
  }

  isEmLimpeza() {
    const s = this._status.toLowerCase();
    return s.includes('limpeza') || s.includes('higieniz') || s.includes('sujo');
  }

  isManutencao() {
    return this._status.toLowerCase().includes('manuten');
  }

  formatarDiaria() {
    return new Intl.NumberFormat('pt-BR', {
      style: 'currency',
      currency: 'BRL'
    }).format(this._diaria);
  }

  /**
   * Ciclo de transição de status para governança e hotelaria
   */
  obterProximoStatusHigienizacao() {
    if (this.isOcupado()) return 'Limpeza Solicitada';
    if (this.isEmLimpeza()) return 'Disponível';
    if (this.isDisponivel()) return 'Em Limpeza';
    return 'Disponível';
  }

  /**
   * Metadados visuais (cor do ponto, sem gerar pill excessiva)
   */
  getStatusMeta() {
    if (this.isDisponivel()) {
      return { label: 'Disponível', dotColor: '#28a745', textColor: '#155724' };
    }
    if (this.isOcupado()) {
      return { label: 'Ocupado', dotColor: '#dc3545', textColor: '#721c24' };
    }
    if (this.isEmLimpeza()) {
      return { label: 'Em Limpeza', dotColor: '#d39e00', textColor: '#856404' };
    }
    return { label: this._status, dotColor: '#6c757d', textColor: '#383d41' };
  }

  toDict() {
    return {
      id: this._id,
      id_quarto: this._id,
      id_hotel: this._idHotel,
      tipo: this._tipo,
      status: this._status,
      andar: this._andar,
      num_quarto: this._numQuarto,
      diaria: this._diaria,
      criado_em: this._criadoEm
    };
  }

  static fromDict(data) {
    if (!data) return null;
    return new Quarto(data);
  }
}
